opnode/rollup/derive/payload_attributes.go

```diff
@@ -236,91 +224,68 @@ func FilterBatches(config *rollup.Config, epoch rollup.Epoch, minL2Time uint64,
 	return
 }
 
-type L2Info interface {
-	Time() uint64
-}
-
-// PayloadAttributes derives a sequence of pre-execution payload attributes from:
-//  - The L1 information the L1-info deposit is derived from
-//  - The L1 receipts the user deposits are derived from
-//  - The data batches the L2 sequencer work is derived from
-//  - The L2 information of the block the new derived blocks build on
-//
-// This is a pure function.
-func PayloadAttributes(config *rollup.Config, l1Info L1Info, receipts []*types.Receipt, seqWindow []*BatchData, l2Info L2Info, minL2Time, maxL2Time uint64) ([]*l2.PayloadAttributes, error) {
-	// Retrieve the deposits of this epoch (all deposits from the first block)
-	deposits, err := DeriveDeposits(l1Info, receipts)
-	if err != nil {
-		return nil, fmt.Errorf("failed to derive deposits: %v", err)
+func ValidBatch(batch *BatchData, config *rollup.Config, epoch rollup.Epoch, minL2Time uint64, maxL2Time uint64) bool {
+	if batch.Epoch != epoch {
+		// Batch was tagged for past or future epoch,
+		// i.e. it was included too late or depends on the given L1 block to be processed first.
+		return false
 	}
-
-	l1Tx := types.NewTx(L1InfoDeposit(l1Info))
-	l1InfoTx, err := l1Tx.MarshalBinary() // TODO: Issue with shared l1InfoTx?
-	if err != nil {
-		return nil, fmt.Errorf("failed to encode L1 info tx")
+	if (batch.Timestamp-config.Genesis.L2Time)%config.BlockTime != 0 {
+		return false // bad timestamp, not a multiple of the block time
 	}
+	if batch.Timestamp < minL2Time {
+		return false // old batch
+	}
+	// limit timestamp upper bound to avoid huge amount of empty blocks
+	if batch.Timestamp >= maxL2Time {
+		return false // too far in future
+	}
+	return true
+}
 
-	// copy L1 randomness (mix-digest becomes randao field post-merge)
-	// TODO: we don't have a randomness oracle on L2, what should sequencing randomness look like.
-	// Repeating the latest randomness of L1 might not be ideal.
-	randomnessSeed := l2.Bytes32(l1Info.MixDigest())
-
-	// Collect all L2 batches, the batches may be out-of-order, or possibly missing.
-	l2Blocks := make(map[uint64]*l2.PayloadAttributes)
-	for _, batch := range seqWindow {
-
-		txns := make([]l2.Data, 0, len(batch.Transactions)+1)
-		txns = append(txns, l1InfoTx)
-		txns = append(txns, batch.Transactions...)
+type L2Info interface {
+	Time() uint64
+}
 
-		l2Blocks[batch.Timestamp] = &l2.PayloadAttributes{
-			Timestamp:             l2.Uint64Quantity(batch.Timestamp),
-			Random:                randomnessSeed,
-			SuggestedFeeRecipient: config.FeeRecipientAddress,
-			Transactions:          txns,
-		}
+// SortedAndPreparedBatches turns a collection of batches to the input batches for a series of blocks
+func SortedAndPreparedBatches(batches []*BatchData, epoch, blockTime, minL2Time, maxL2Time uint64) []*BatchData {
+	m := make(map[uint64]*BatchData)
+	for _, b := range batches {
+		b := b // not sure if copy is required, but go does not make this easy
```

Comment:
Should not be required. We're only dealing with a pointer here, we already copy the pointer when we set `m[b.BatchV1.Timestamp]` below.

Reply:
Right yes. Double checked and the problem is only when taking the address of the loop variable.